File: hot100/Solution0234.cpp
using namespace std;
// 给你一个单链表的头节点 head ，请你判断该链表是否为回文链表。如果是，返回 true ；否则，返回 false 。

// 示例 1：

// 输入：head = [1,2,2,1]
// 输出：true
// 示例 2：

// 输入：head = [1,2]
// 输出：false

// 提示：

// 链表中节点数目在范围[1, 105] 内
// 0 <= Node.val <= 9

// 进阶：你能否用 O(n) 时间复杂度和 O(1) 空间复杂度解决此题？
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode() : val(0), next(nullptr) {}
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

    ListNode *reverseList(ListNode *head)
    {
        ListNode *prev = nullptr;
        ListNode *curr = head;
        while (curr)
        {
            ListNode *next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

public:
    bool isPalindrome(ListNode *head)
    {
        if (!head || !head->next)
            return true; // 空链表或单节点链表是回文
        ListNode *slow_pointer = head, *fast_pointer = head;
        while (fast_pointer->next && fast_pointer->next->next)
        {
            fast_pointer = fast_pointer->next->next;
            slow_pointer = slow_pointer->next;
        }
        // 如果链表长度是奇数，slow_pointer是中间节点，fast_pointer是最后一个节点
        // 如果链表长度是偶数，slow_pointer是前半部分最后一个节点，fast_pointer是倒数第二个节点
        // 因此无论是奇数还是偶数，slow_pointer的下一个节点就是后半部分的第一个节点
        ListNode *secondHalf = reverseList(slow_pointer->next);
        ListNode *secondHalfHead = secondHalf; // 保存用于恢复

        bool result = true;
        ListNode *firstHalf = head;
        while (secondHalf)
        {
            if (firstHalf->val != secondHalf->val)
            {
                result = false;
                break;
            }
            firstHalf = firstHalf->next;
            secondHalf = secondHalf->next;
        }
        slow_pointer->next = reverseList(secondHalfHead); // 恢复链表
        return result;
    }
};

// 方法二：递归。空间复杂度为O(n)
class Solution2
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode() : val(0), next(nullptr) {}
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };
    ListNode *front_pointer;

    bool recursivelyCheck(ListNode *current)
    {
        if (current)
        {
            if (!recursivelyCheck(current->next))
            {
                return false;
            }
            if (current->val != front_pointer->val)
            {
                return false;
            }
            front_pointer = front_pointer->next;
        }
        return true;
    }

public:
    bool isPalindrome(ListNode *head)
    {
        front_pointer = head;
        return recursivelyCheck(head);
    }
};

File: hot100/Solution0234.java
package hot100;

// 给你一个单链表的头节点 head ，请你判断该链表是否为回文链表。如果是，返回 true ；否则，返回 false 
public class Solution0234 {
    private static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode secondHalfHead = slow.next;
        ListNode reversedSecondHalfHead = reverseList(secondHalfHead);
        ListNode temp = reversedSecondHalfHead;
        ListNode firstHalfHead = head;
        boolean isPalindrome = true;
        while (reversedSecondHalfHead != null) {
            if (firstHalfHead.val != reversedSecondHalfHead.val) {
                isPalindrome = false;
                break;
            }
            firstHalfHead = firstHalfHead.next;
            reversedSecondHalfHead = reversedSecondHalfHead.next;
        }
        slow.next = reverseList(temp);
        return isPalindrome;
    }

    private ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    private ListNode frontNode = null;

    public boolean isPalindrome2(ListNode head) {
        frontNode = head;
        return recursivelyCheck(head);
    }

    private boolean recursivelyCheck(ListNode currentNode) {
        if (currentNode != null) {
            if (!recursivelyCheck(currentNode.next)) {
                return false;
            }
            if (currentNode.val != frontNode.val) { // 奇数递归时，currentNode和frontNode会指向同一个节点；偶数递归到这里，currentNode和frontNode会指向对半的两个节点，递归出栈后，分别向两边移动，直到相遇
                return false;
            }
            frontNode = frontNode.next;
        }
        return true;
    }

    public static void main(String[] args) {
        Solution0234 solution = new Solution0234();
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(2);
        head.next.next.next = new ListNode(1);

        boolean result = solution.isPalindrome(head);
        System.out.println("链表是否为回文链表: " + result);
        boolean result2 = solution.isPalindrome2(head);
        System.out.println("链表是否为回文链表(递归方法): " + result2);
    }

}
